lib/SPIRV/SPIRVToOCL.cpp

```diff
@@ -0,0 +1,470 @@
+      std::swap(Args[ScopeIdx], Args.back());
+      if (OC == OpAtomicCompareExchange ||
+        OC == OpAtomicCompareExchangeWeak) {
+        // OpAtomicCompareExchange[Weak] semantics is different from
+        // atomic_compare_exchange_[strong|weak] semantics as well as
+        // arguments order.
+        // OCL built-ins returns boolean value and stores a new/original
+        // value by pointer passed as 2nd argument (aka expected) while SPIR-V
+        // instructions returns this new/original value as a resulting value.
+        AllocaInst *PExpected =
+          new AllocaInst(CI->getType(), 0, "expected",
+            &(*PInsertBefore->getParent()
+              ->getParent()
+              ->getEntryBlock()
+              .getFirstInsertionPt()));
+        PExpected->setAlignment(CI->getType()->getScalarSizeInBits() / 8);
+        new StoreInst(Args[1], PExpected, PInsertBefore);
+        unsigned AddrSpc = SPIRAS_Generic;
+        Type *PtrTyAS =
+          PExpected->getType()->getElementType()->getPointerTo(AddrSpc);
+        Args[1] = CastInst::CreatePointerBitCastOrAddrSpaceCast(
+          PExpected, PtrTyAS, PExpected->getName() + ".as", PInsertBefore);
+        std::swap(Args[3], Args[4]);
+        std::swap(Args[2], Args[3]);
+        RetTy = Type::getInt1Ty(*Ctx);
+      }
+      return Name;
+    },
+    [=](CallInst *CI) -> Instruction * {
+      if (OC == OpAtomicCompareExchange ||
+          OC == OpAtomicCompareExchangeWeak) {
+        // OCL built-ins atomic_compare_exchange_[strong|weak] return boolean
+        // value. So, to obtain the same value as SPIR-V instruction is
+        // returning it has to be loaded from the memory where 'expected'
+        // value is stored. This memory must contain the needed value after a
+        // call to OCL built-in is completed.
+        LoadInst *POriginal =
+          new LoadInst(CI->getArgOperand(1), "original", PInsertBefore);
+        return POriginal;
+      }
+      // For other built-ins the return values match.
+      return CI;
+    },
+    &Attrs);
+}
+
+} // namespace SPIRV
+
+ModulePass *llvm::createSPIRVToOCL(Module &M) {
+  if (OCLBuiltinsVersion.getNumOccurrences() > 0) {
+    if (OCLBuiltinsVersion.getValue() == 12)
+      return createSPIRVToOCL12();
+    else if (OCLBuiltinsVersion.getValue() == 20 ||
+      OCLBuiltinsVersion.getValue() == 21)
+      return createSPIRVToOCL20();
+    else
+      assert("Invalid spirv-ocl-builtins-version value");
+  }
+  // Below part of code is here just temporarily (to not broke existing
+  // projects based on translator), because ocl builtins versions shouldn't has a dependency
+  // on OpSource spirv opcode. 
+  // OpSource spec: "This has no semantic impact and can safely be removed from a module."
+  // After some time it can be removed, then only factor impacting version of ocl builtins
+  // will be spirv-ocl-builtins-version command option.
+  unsigned OCLVersion = getOCLVersion(&M);
```

Comment:
Please, pay attention on this part of code.
I've added an option to specify target ocl builtins version, but I left the logic based on OpSource opcode as it was working before to avoid crashing the projects that are dependent on the Translator, but don't have an option specified yet. Is it a good approach, or we should remove that logic permanently in this commit?

Reply:
It sounds reasonable to have a transition period.  I wonder if we should emit a warning that this behaviour is now deprecated?

Reply:
Right, do we have any specific mechanism that could let us to generate such warning?

Reply:
We have mechanism to emit errors, but I'm afraid that we don't have anything for warnings.

At least, we can document this change somewhere in README or in the special file which describes translation from SPIR-V to LLVM IR for OpenCL

Reply:
Added short info about new option to README. Don't know if it's enough.